File: agentic_os/mission/verify.py
"""P7B — the Verification Plane: acceptance control for consequential state transitions.

The invariant: *no consequential operator result becomes authoritative World State until the
applicable acceptance policy succeeds.* A result runs the pipeline

    Operator Result → syntactic → policy → state-transition (assertions)
                    → ACCEPT   —or—   REJECT (hard) / ESCALATE (human decides)

Verification is a runtime plane, not merely a node type: the compiler may attach assertions in
advance, and the runtime invokes the verifier before accepting any consequential transition.
"""
from __future__ import annotations

from typing import Callable, Protocol

from .registry import CapabilityRegistry
from .types import StateAssertion, VerificationDecision, VerificationResult


class Verifier(Protocol):
    def verify(self, node, result, world, mission) -> VerificationResult: ...


def needs_verification(node) -> bool:
    """Consequential = a side effect, or a node carrying explicit assertions to check."""
    return bool(getattr(node, "side_effecting", False) or getattr(node, "assertions", None))


class CompositeVerifier:
    """Default acceptance pipeline. Hard failures (malformed result / lost permission) REJECT;
    unmet assertions ESCALATE to a human; otherwise ACCEPT."""

    def __init__(self, registry: CapabilityRegistry, granted_of: Callable[[object], set]):
        self.registry = registry
        self.granted_of = granted_of

    def verify(self, node, result, world, mission) -> VerificationResult:
        checks: list[dict] = []
        cap = self.registry.get(node.capability)

        # 1. syntactic — the result carries the capability's declared outputs
        outs = list((cap.outputs or {}).keys()) if cap else []
        syn_ok = isinstance(result, dict) and all(k in result for k in outs)
        checks.append({"stage": "syntactic", "passed": syn_ok, "detail": f"expects outputs {outs}"})

        # 2. policy — the capability's permissions still hold under the mission's grants
        granted = self.granted_of(mission)
        pol_ok = "*" in granted or cap is None or all(p in granted for p in cap.permissions)
        checks.append({"stage": "policy", "passed": pol_ok, "detail": "permissions still granted"})

        # 3. state-transition — every attached assertion holds against the result / world
        unmet = []
        for a in getattr(node, "assertions", []) or []:
            holds = _assert_holds(a, result, world)
            checks.append({"stage": "assertion", "passed": holds,
                           "detail": f"{a.key} {a.op} {a.expected!r}"})
            if not holds:
                unmet.append(a.key)

        if not syn_ok or not pol_ok:
            decision = VerificationDecision.REJECT
        elif unmet:
            decision = VerificationDecision.ESCALATE
        else:
            decision = VerificationDecision.ACCEPT
        conf = sum(1 for c in checks if c["passed"]) / max(1, len(checks))
        return VerificationResult(decision=decision, confidence=round(conf, 3),
                                  checks=checks, verifier="composite")


def _assert_holds(a: StateAssertion, result, world) -> bool:
    val = result.get(a.key) if isinstance(result, dict) and a.key in result else (
        world.get(a.key) if world is not None else None)
    a.observed = val
    if a.op == "truthy":
        a.holds = bool(val)
    elif a.op == "equals":
        a.holds = val == a.expected
    elif a.op == "in":
        a.holds = val in (a.expected or [])
    else:  # present
        a.holds = val is not None
    return bool(a.holds)

File: go/mission/verify.go
package mission

// P7B — the Verification Plane: acceptance control for consequential state transitions. The
// invariant: no consequential operator result becomes authoritative World State until the
// applicable acceptance policy succeeds. Pipeline: syntactic → policy → state-transition →
// ACCEPT / REJECT (hard) / ESCALATE (human decides). Go port of agentic_os/mission/verify.py.

import (
	"fmt"
	"math"
	"reflect"
)

// Verifier accepts, rejects or escalates a consequential result.
type Verifier interface {
	Verify(node *Node, result map[string]any, world *WorldState, mission *Mission) VerificationResult
}

// needsVerification: consequential = a side effect, or a node carrying explicit assertions.
func needsVerification(node *Node) bool {
	return node.SideEffecting || len(node.Assertions) > 0
}

// CompositeVerifier is the default acceptance pipeline. Hard failures (malformed result / lost
// permission) REJECT; unmet assertions ESCALATE; otherwise ACCEPT.
type CompositeVerifier struct {
	registry  Registry
	grantedOf func(*Mission) map[string]bool
}

func NewCompositeVerifier(registry Registry, grantedOf func(*Mission) map[string]bool) *CompositeVerifier {
	return &CompositeVerifier{registry: registry, grantedOf: grantedOf}
}

func (v *CompositeVerifier) Verify(node *Node, result map[string]any, world *WorldState, mission *Mission) VerificationResult {
	var checks []map[string]any
	cap := v.registry.Get(node.Capability)

	// 1. syntactic — the result carries the capability's declared outputs
	var outs []string
	if cap != nil {
		outs = mapKeys(cap.Outputs)
	}
	synOK := true
	for _, k := range outs {
		if _, ok := result[k]; !ok {
			synOK = false
			break
		}
	}
	checks = append(checks, map[string]any{"stage": "syntactic", "passed": synOK, "detail": fmt.Sprintf("expects outputs %v", outs)})

	// 2. policy — the capability's permissions still hold under the mission's grants
	granted := v.grantedOf(mission)
	polOK := granted["*"] || cap == nil
	if !polOK {
		polOK = true
		for _, p := range cap.Permissions {
			if !granted[p] {
				polOK = false
				break
			}
		}
	}
	checks = append(checks, map[string]any{"stage": "policy", "passed": polOK, "detail": "permissions still granted"})

	// 3. state-transition — every attached assertion holds against the result / world
	var unmet []string
	for i := range node.Assertions {
		a := &node.Assertions[i]
		holds := assertHolds(a, result, world)
		checks = append(checks, map[string]any{"stage": "assertion", "passed": holds,
			"detail": fmt.Sprintf("%s %s %v", a.Key, a.Op, a.Expected)})
		if !holds {
			unmet = append(unmet, a.Key)
		}
	}

	var decision VerificationDecision
	switch {
	case !synOK || !polOK:
		decision = VerReject
	case len(unmet) > 0:
		decision = VerEscalate
	default:
		decision = VerAccept
	}
	passed := 0
	for _, c := range checks {
		if c["passed"] == true {
			passed++
		}
	}
	conf := float64(passed) / math.Max(1, float64(len(checks)))
	vr := NewVerificationResult(decision)
	vr.Confidence = roundN(conf, 3)
	vr.Checks = checks
	return vr
}

func assertHolds(a *StateAssertion, result map[string]any, world *WorldState) bool {
	var val any
	if result != nil {
		if v, ok := result[a.Key]; ok {
			val = v
		} else if world != nil {
			val = world.Get(a.Key)
		}
	} else if world != nil {
		val = world.Get(a.Key)
	}
	a.Observed = val
	var holds bool
	switch a.Op {
	case "truthy":
		holds = truthy(val)
	case "equals":
		holds = reflect.DeepEqual(val, a.Expected)
	case "in":
		holds = inList(val, a.Expected)
	default: // present
		holds = val != nil
	}
	a.Holds = &holds
	return holds
}

func inList(val, expected any) bool {
	switch xs := expected.(type) {
	case []any:
		for _, x := range xs {
			if reflect.DeepEqual(x, val) {
				return true
			}
		}
	case []string:
		if s, ok := val.(string); ok {
			return strContains(xs, s)
		}
	}
	return false
}
